### PR TITLE
Improve user management section

### PR DESCRIPTION
**Pull request by DonOmalVindula**
### Purpose
$subject

### Related Issues
- https://github.com/wso2/product-is/issues/17831

### Related PRs
- Related PR `#1` or (None)

### Checklist
- [ ] e2e cypress tests locally verified.
- [ ] Manual test round performed and verified.
- [ ] UX/UI review done on the final implementation.
- [ ] Documentation provided. (Add links if there are any)
- [ ] Unit tests provided. (Add links if there are any)
- [ ] Integration tests provided. (Add links if there are any)

### Security checks
- [ ] Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines?
- [ ] Ran FindSecurityBugs plugin and verified report?
- [ ] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets?

### CODE REVIEW

File: apps/console/src/features/users/api/invite.ts [hunk 1]
```diff
@@ -0,0 +1,213 @@
+/**
+ * Copyright (c) 2021, WSO2 LLC. (https://www.wso2.com).
```
Comment (Achintha444) [inline apps/console/src/features/users/api/invite.ts:2]:
```suggestion
 * Copyright (c) 2023, WSO2 LLC. (https://www.wso2.com).
```

**Review by Achintha444 — APPROVED** [review]

File: apps/console/src/features/users/api/invite.ts [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+import { Config } from "../../core/configs";
+import useRequest, { RequestErrorInterface, RequestResultInterface } from "../../core/hooks/use-request";
+import { store } from "../../core/store";
+import { UserManagementConstants } from "../constants";
+import { UserInviteInterface } from "../models";
+
+/**
+ * Initialize an axios Http client.
+ *
+ */
```
Comment (pavinduLakshan) [inline apps/console/src/features/users/api/invite.ts:32]: Do we really need this comment?

File: apps/console/src/features/users/api/invite.ts [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    const requestConfig: AxiosRequestConfig = {
+        headers: {
+            "Access-Control-Allow-Origin": store.getState().config.deployment.clientHost,
+            "Content-Type": "application/json"
+        },
+        method: HttpMethods.GET,
+        url: store.getState().config.endpoints.inviteEndpoint
+    };
+
+    return httpClient(requestConfig).then((response: AxiosResponse) => {
```
Comment (pavinduLakshan) [inline apps/console/src/features/users/api/invite.ts:46]: Any reason not to use SWR here?

PS: It seems like this function is kept for compatibility. If that's the case, let's mark this function as deprecated :)

File: apps/console/src/features/users/api/invite.ts [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    return {
+        data,
+        error: error,
+        isLoading: !error && !data,
+        isValidating,
+        mutate: mutate
+    };
+};
+
+export const sendInvite = (userInvite: UserInviteInterface): Promise<any> => {
```
Comment (pavinduLakshan) [inline apps/console/src/features/users/api/invite.ts:82]: Let's add doc comments. Please check the other places as well.

File: apps/console/src/features/users/api/invite.ts [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        headers: {
+            "Access-Control-Allow-Origin": store.getState().config.deployment.clientHost,
+            "Content-Type": "application/json"
+        },
+        method: HttpMethods.POST,
+        url: store.getState().config.endpoints.inviteEndpoint
+    };
+
+    return httpClient(requestConfig).then((response: AxiosResponse) => {
+        return Promise.resolve(response);
```
Comment (pavinduLakshan) [inline apps/console/src/features/users/api/invite.ts:94]: No need to wrap again in Promise.resolve as httpClient already returns the value as a promise

File: apps/console/src/features/users/api/invite.ts [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    }).catch((error: AxiosError) => {
+        return Promise.reject(error);
+    });
+};
+
+export const deleteInvite = (traceID: string): Promise<any> => {
```
Comment (pavinduLakshan) [inline apps/console/src/features/users/api/invite.ts:100]: WDYT?

```suggestion
export const deleteInvite = (inviteID: string): Promise<any> => {
```

File: apps/console/src/features/users/components/add-user-updated.tsx [hunk 7]
```diff
@@ -0,0 +1,947 @@
+/**
+ * Copyright (c) 2021, WSO2 LLC. (https://www.wso2.com).
```
Comment (pavinduLakshan) [inline apps/console/src/features/users/components/add-user-updated.tsx:2]:
```suggestion
 * Copyright (c) 2023, WSO2 LLC. (https://www.wso2.com).
```

File: apps/console/src/features/users/components/add-user-updated.tsx [hunk 8]
```diff
@@ -0,0 +1,947 @@
...
+/**
+ * Proptypes for the add user component.
+ */
+export interface AddUserProps {
+    initialValues: any;
+    triggerSubmit: boolean;
+    emailVerificationEnabled: boolean;
+    onSubmit: (values: BasicUserDetailsInterface) => void;
+    hiddenFields?: ("userStore" |"userName" | "firstName" | "lastName" | "password" | "email")[];
+    requestedPasswordOption?: "ask-password" | "create-password";
```
Comment (pavinduLakshan) [inline apps/console/src/features/users/components/add-user-updated.tsx:50]: Shall we create an enum for this?

File: apps/console/src/features/users/components/add-user-updated.tsx [hunk 9]
```diff
@@ -0,0 +1,947 @@
...
+    const [ userstore, setUserstore ] = useState<string>(PRIMARY_USERSTORE);
+    const [ isValidEmail, setIsValidEmail ] = useState<boolean>(false);
+    const [ isEmailFilled, setIsEmailFilled ] = useState<boolean>(false);
+
+    const formBottomRef: MutableRefObject<HTMLDivElement> = useRef<HTMLDivElement>();
+    const emailRef: MutableRefObject<HTMLDivElement> = useRef<HTMLDivElement>();
+
+    const { t } = useTranslation();
+
+    const [ , setRegExLoading ] = useState<boolean>(false);
```
Comment (pavinduLakshan) [inline apps/console/src/features/users/components/add-user-updated.tsx:113]: We can remove this, can't we?

File: apps/console/src/features/users/components/add-user-updated.tsx [hunk 10]
```diff
@@ -0,0 +1,947 @@
...
+            setUsernameConfig(getUsernameConfiguration(validationConfig));
+        }
+    }, [ validationConfig ]);
+
+    useEffect(() => {
+        if (!isAlphanumericUsernameEnabled()) {
+            setIsValidEmail(true);
+        } else {
+            setIsValidEmail(false);
+            // setAskPasswordOption("offline");
```
Comment (pavinduLakshan) [inline apps/console/src/features/users/components/add-user-updated.tsx:178]: Let's remove this if unused

File: apps/console/src/features/users/components/add-user-updated.tsx [hunk 11]
```diff
@@ -0,0 +1,947 @@
...
+        }
+    }, [ usernameConfig ]);
+
+    /**
+     * Check whether the alphanumeric usernames are enabled.
+     *
+     * @returns isAlphanumericUsernameEnabled - validation status.
+     */
+    const isAlphanumericUsernameEnabled= (): boolean => {
+
```
Comment (pavinduLakshan) [inline apps/console/src/features/users/components/add-user-updated.tsx:188]:
```suggestion
```

File: apps/console/src/features/users/components/add-user-updated.tsx [hunk 12]
```diff
@@ -0,0 +1,947 @@
...
+    };
+
+    /**
+     * Callback function to validate password.
+     *
+     * @param valid - validation status.
+     * @param validationStatus - detailed validation status.
+     */
+    const onPasswordValidate = (valid: boolean): void => {
+
```
Comment (pavinduLakshan) [inline apps/console/src/features/users/components/add-user-updated.tsx:199]:
```suggestion
```

File: apps/console/src/features/users/components/add-user-updated.tsx [hunk 13]
```diff
@@ -0,0 +1,947 @@
...
+        setRegExLoading(true);
+        await SharedUserStoreUtils.getUserStoreRegEx(userstore,
+            SharedUserStoreConstants.USERSTORE_REGEX_PROPERTIES.UsernameRegEx)
+            .then((response: string) => {
+                setUserStoreRegex(response);
+                setRegExLoading(false);
+            });
+    };
+
+    useEffect(() => {
```
Comment (pavinduLakshan) [inline apps/console/src/features/users/components/add-user-updated.tsx:216]: Let's group all useEffect s to one place

File: apps/console/src/features/users/components/add-user-updated.tsx [hunk 14]
```diff
@@ -0,0 +1,947 @@
...
+     * The following function handles email empty validation.
+     *
+     * @param values - Map of form values.
+     */
+    const handleEmailEmpty = (values: Map<string, FormValue>): void => {
+        if (values.get("email")?.toString() === "") {
+            setIsEmailFilled(false);
+        } else {
+            setIsEmailFilled(true);
+        }
```
Comment (pavinduLakshan) [inline apps/console/src/features/users/components/add-user-updated.tsx:253]:
```suggestion
        return values.get("email")?.toString() === ""
```

File: apps/console/src/features/users/components/add-user-updated.tsx [hunk 15]
```diff
@@ -0,0 +1,947 @@
...
+    };
+
+    /**
+     * Verify whether the provided password is valid.
+     *
+     * @param password - The password to validate.
+     */
+    const isNewPasswordValid = async (password: string) => {
+
```
Comment (pavinduLakshan) [inline apps/console/src/features/users/components/add-user-updated.tsx:303]:
```suggestion
```

File: apps/console/src/features/users/api/invite.ts (hunk 2 above)
Comment (DonOmalVindula) [reply apps/console/src/features/users/api/invite.ts]: Fixed with fd5839362baf8e4e50397013a20cfa8b0b15fa4e

File: apps/console/src/features/users/api/invite.ts (hunk 3 above)
Comment (DonOmalVindula) [reply apps/console/src/features/users/api/invite.ts]: Fixed with fd5839362baf8e4e50397013a20cfa8b0b15fa4e

File: apps/console/src/features/users/api/invite.ts (hunk 4 above)
Comment (DonOmalVindula) [reply apps/console/src/features/users/api/invite.ts]: Fixed with fd5839362baf8e4e50397013a20cfa8b0b15fa4e

File: apps/console/src/features/users/api/invite.ts (hunk 5 above)
Comment (DonOmalVindula) [reply apps/console/src/features/users/api/invite.ts]: Fixed with fd5839362baf8e4e50397013a20cfa8b0b15fa4e

File: apps/console/src/features/users/api/invite.ts (hunk 6 above)
Comment (DonOmalVindula) [reply apps/console/src/features/users/api/invite.ts]: Fixed with fd5839362baf8e4e50397013a20cfa8b0b15fa4e

File: apps/console/src/features/users/components/add-user-updated.tsx (hunk 7 above)
Comment (DonOmalVindula) [reply apps/console/src/features/users/components/add-user-updated.tsx]: Fixed with fd5839362baf8e4e50397013a20cfa8b0b15fa4e

File: apps/console/src/features/users/components/add-user-updated.tsx (hunk 9 above)
Comment (DonOmalVindula) [reply apps/console/src/features/users/components/add-user-updated.tsx]: Fixed with fd5839362baf8e4e50397013a20cfa8b0b15fa4e

File: apps/console/src/features/users/components/add-user-updated.tsx (hunk 10 above)
Comment (DonOmalVindula) [reply apps/console/src/features/users/components/add-user-updated.tsx]: Fixed with fd5839362baf8e4e50397013a20cfa8b0b15fa4e

File: apps/console/src/features/users/components/add-user-updated.tsx (hunk 11 above)
Comment (DonOmalVindula) [reply apps/console/src/features/users/components/add-user-updated.tsx]: Fixed with fd5839362baf8e4e50397013a20cfa8b0b15fa4e

File: apps/console/src/features/users/components/add-user-updated.tsx (hunk 12 above)
Comment (DonOmalVindula) [reply apps/console/src/features/users/components/add-user-updated.tsx]: Fixed with fd5839362baf8e4e50397013a20cfa8b0b15fa4e

File: apps/console/src/features/users/components/add-user-updated.tsx (hunk 13 above)
Comment (DonOmalVindula) [reply apps/console/src/features/users/components/add-user-updated.tsx]: Fixed with fd5839362baf8e4e50397013a20cfa8b0b15fa4e

File: apps/console/src/features/users/components/add-user-updated.tsx (hunk 14 above)
Comment (DonOmalVindula) [reply apps/console/src/features/users/components/add-user-updated.tsx]: Fixed with fd5839362baf8e4e50397013a20cfa8b0b15fa4e

File: apps/console/src/features/users/components/add-user-updated.tsx (hunk 15 above)
Comment (DonOmalVindula) [reply apps/console/src/features/users/components/add-user-updated.tsx]: Fixed with fd5839362baf8e4e50397013a20cfa8b0b15fa4e

File: apps/console/src/features/users/components/add-user-updated.tsx (hunk 8 above)
Comment (DonOmalVindula) [reply apps/console/src/features/users/components/add-user-updated.tsx]: Fixed with fd5839362baf8e4e50397013a20cfa8b0b15fa4e

**Review by pavinduLakshan — DISMISSED** [review]

**Review by Achintha444 — APPROVED** [review]

**Pull request merged**